### DISCUSSION TITLE
Create a middleware to sanitize the request body

### DISCUSSION THREAD

**Discussion by livingdotone** [fastapi/fastapi#8354] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import Message
from fastapi import Request
import bleach
import json


class SanitizeMiddleware(BaseHTTPMiddleware):

    def __init__(self, app):
        super().__init__(app)

    @staticmethod
    def __sanitize_array(array_values):
        for index, value in enumerate(array_values):
            if isinstance(value, dict):
                array_values[index] = {key: bleach.clean(
                    value) for key, value in value.items()}
            else:
                array_values[index] = bleach.clean(value)
        return array_values

    async def set_body(self, request: Request):
        receive_ = await request._receive()
        async def receive() -> Message:
            return receive_

        request._receive = receive

    async def dispatch(self, request: Request, call_next):
        await self.set_body(request)

        if request.method == 'POST' or request.method == 'PUT':
            json_body = await request.json()
            sanitize_body = {key: self.__sanitize_array(value) if isinstance(
                value, list) else bleach.clean(value) for key, value in json_body.items()}
            request._body = json.dumps(
                sanitize_body, indent=2).encode('utf-8')
            await self.set_body(request)

        response = await call_next(request)
        return response
```

### Description

Hello guys, I'm facing an issue while trying to sanitize the request body from POST and PUT methods. The idea is to sanitize the request.body() before it even reachs the view. For instance, I would like to pass _bleach_ on it to avoid security issues that might appear under the body that is sent.
I've built a middleware that inherit from BaseHTTPMiddleware, to sanetize the body, but I've notived that I'm not changing the original _request_ element:

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.8.10

**Comment by JarroVGIT** [comment]
You never return your request from `set_body`. Try this:

```python
class SanitizeMiddleware(BaseHTTPMiddleware):

    def __init__(self, app):
        super().__init__(app)

    @staticmethod
    def __sanitize_array(array_values):
        for index, value in enumerate(array_values):
            if isinstance(value, dict):
                array_values[index] = {key: bleach.clean(
                    value) for key, value in value.items()}
            else:
                array_values[index] = bleach.clean(value)
        return array_values

    async def set_body(self, request: Request):
        receive_ = await request._receive()
        async def receive() -> Message:
            return receive_

        request._receive = receive
        return request

    
    async def dispatch(self, request: Request, call_next):
        await self.set_body(request)

        if request.method == 'POST' or request.method == 'PUT':
            json_body = await request.json()
            sanitize_body = {key: self.__sanitize_array(value) if isinstance(
                value, list) else bleach.clean(value) for key, value in json_body.items()}
            request._body = json.dumps(
                sanitize_body, indent=2).encode('utf-8')
            request = await self.set_body(request)

        response = await call_next(request)
        return response
```

**Comment by livingdotone** [comment]
> You never return your request from `set_body`. Try this:
> 
> ```python
> class SanitizeMiddleware(BaseHTTPMiddleware):
> 
>     def __init__(self, app):
>         super().__init__(app)
> 
>     @staticmethod
>     def __sanitize_array(array_values):
>         for index, value in enumerate(array_values):
>             if isinstance(value, dict):
>                 array_values[index] = {key: bleach.clean(
>                     value) for key, value in value.items()}
>             else:
>                 array_values[index] = bleach.clean(value)
>         return array_values
> 
>     async def set_body(self, request: Request):
>         receive_ = await request._receive()
>         async def receive() -> Message:
>             return receive_
> 
>         request._receive = receive
>         return request
> 
>     
>     async def dispatch(self, request: Request, call_next):
>         await self.set_body(request)
> 
>         if request.method == 'POST' or request.method == 'PUT':
>             json_body = await request.json()
>             sanitize_body = {key: self.__sanitize_array(value) if isinstance(
>                 value, list) else bleach.clean(value) for key, value in json_body.items()}
>             request._body = json.dumps(
>                 sanitize_body, indent=2).encode('utf-8')
>             request = await self.set_body(request)
> 
>         response = await call_next(request)
>         return response
> ```


I tested it here and it continues with the same previous request

**Comment by JarroVGIT** [comment]
Hmm I did not expect that.. 🤔 I am tinkering around with this and I notice that there is a custom request being send to the next ASGI app, but the body is unsanitized. If I include a custom key in the `request.scope`, then that key (and value) is available in my endpoint. 

Somewhere, Starlette (or FastAPI but I am thinking Starlette) is unsanitizing the body. Which is very weird!

**Comment by MarcelFox** [comment]
I think that maybe the request element isn't the same at all, I've noticed that the _id_ for both are different, not sure if that tells much but I'm assuming that they might be addressing to different elements in the memory, but is pure speculation at this point.

**Comment by xingdongzhe** [comment]
@MarcelFox You are right for this
> I've noticed that the id for both are different

In `BaseHTTPMiddleware`, It will be new a request instance pass to next middleware or route endpoint
At this point, we only need the same request instance, so @casluxd you can do this two ways:
- pass request to scope: `scope["request"]=sanitized_reqeust`
- use contextvars: `sanitized_request_context=contextvars.Contexvars("request"), sanitized_request_context.set(sanitized_reqeust)`  

So you can get middlerware request in your route endpoint

**Reply by Balu-enp** [reply]
Have you guys fixed the issue. If fixed Please provide the fixed code

**Reply by cemsozens** [reply]
Any updates for this?

**Comment by cemsozens** [comment, marked as answer]
Use Pure ASGI Middleware instead to achieve this @casluxd. You can see that this overcomes the limitations of base middleware. https://www.starlette.io/middleware/

**Comment by cemsozens** [comment]
```
import json

import bleach
from starlette.types import ASGIApp, Scope, Receive, Send
import functools
import typing


def __sanitize_array(array_values: list[typing.Any]) -> list[typing.Any]:
    for index, value in enumerate(array_values):
        if isinstance(value, dict):
            array_values[index] = {key: bleach.clean(
                value) if isinstance(value, str) else value for key, value in value.items()}
        else:
            array_values[index] = bleach.clean(value) if isinstance(value, str) else value
    return array_values


def sanitize(func: Receive) -> Receive:
    @functools.wraps(func)
    async def wrapper(*args: typing.Any, **kwargs: typing.Any) -> typing.MutableMapping[str, typing.Any]:
        message = await func(*args, **kwargs)
        body = message.get("body")
        if not body:
            return message
        if not isinstance(body, bytes) :
            return message
        json_body = json.loads(body)
        for key, value in json_body.items():
            if isinstance(value, dict):
                json_body[key] = {key: bleach.clean(
                    value) if isinstance(value, str) else value for key, value in value.items()}
            elif isinstance(value, list):
                json_body[key] = __sanitize_array(value)
            else:
                json_body[key] = bleach.clean(value)
        message["body"] = bytes(json.dumps(json_body), encoding="utf-8")
        return message
    return wrapper


class SanitizeMiddleware:
    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if "method" not in scope or scope["method"] in ("GET", "HEAD"):
            await self.app(scope, receive, send)
            return
        await self.app(scope, sanitize(receive), send)
```

then you can add it like 
```
api = fastapi.FastAPI(
    middleware=[Middleware(SanitizeMiddleware)],
)
```

**Reply by cemsozens** [reply]
Note: I will try to create a pr for new middleware in starlette.

**Reply by MarcelFox** [reply]
That's a great idea! Can you send the PR link afterwards?

**Reply by Kludex** [reply]
> Note: I will try to create a pr for new middleware in starlette.

To do what?

**Reply by cemsozens** [reply]
@Kludex to sanitize the body. Not everyone needs to add as middleware though of course.

**Reply by Francis-Ani** [reply]
This works great! Note you also need to check the value is a string in the else block of the `wrapper` function like in the other spots.
`json_body[key] = bleach.clean(value) if isinstance(value, str) else value`
